bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdf_pyz.py

```diff
@@ -0,0 +1,291 @@
+            if idx < n_cols:
+                self.func_args[p] = cols_list[idx]
+        self.func_args = {**self.func_args, **extra_args} # Extra args supersedes col_list
+
+    def find_function_signature(self):
+        """
+        Calculates the function signature. 
+        Updates the class with a new attribute args_info which contains the mapping of the parameter of the function to its type and value.
+        args_info:  dict{parameter_name(str): (type, value)}
+        """
+        func_args = self.func_args
+        for p in self.params:
+            if p in func_args: # the parameter value has been given in func_args
+                value_of_p = func_args[p]
+                type_of_p = self.find_type(value_of_p)
+                if type(value_of_p) == bool: # Bool(s) in python are represented as True/False but in C++ are true/false. The following if statements are to account for that
+                    if value_of_p: value_of_p = 'true'
+                    else: value_of_p = 'false'
+            else: # the parameter was not in func_args. Thus this parameter has to be mapped to a column of rdf
+                if p not in self.col_names:
+                    raise Exception(f"Unable to map function argument {p} to a column.\nUse correct name of column or pass a list of column names.")
+                value_of_p = p 
+                type_of_p = self.find_type(p)
+            self.args_info[p] = (type_of_p, value_of_p)
+
+    def generate_function_call(self):
+        """
+        Generates a function call and signature. Gives a unique name to the function if it is a lambda function.
+        Updates the class with new attributes func_call and func_sign to hold them,
+        """
+        func = self.func
+        if func.__name__ == '<lambda>' : 
+            func.__name__ = f"_lambda_func_number_{FunctionJitter.lambda_function_counter}"
+            FunctionJitter.lambda_function_counter += 1
+        self.func_call = f"{func.__name__}({', '.join(str(arg_info[1]) for arg_info in self.args_info.values())})"
+        self.func_sign = [str(self.args_info[p][0]) for p in self.params]
+
+            
+    def jit_function(self, func, cols_list, extra_args):
+        """
+        Jits the provided function using ROOT's NumbaDeclare.
+        Also checks if the function was jitted earlier in which case it won't jit again but if signature does not match. It raises an error.
+
+        Arguments:
+        func: A python callable
+        cols_list: A list of columns of RDF on which func depends on.
+        extra_args: A dict of extra arguments that func requires.
+        """
+        if func.__name__ in FunctionJitter.function_cache:
+            func_call, func_sign = FunctionJitter.function_cache[func.__name__]
+            self.find_function_params(func)
+            self.generate_func_args(cols_list, extra_args)
+            self.find_function_signature()
+            self.generate_function_call()
```

Comment:
If these four method calls need to be performed in that order (there are hidden dependencies between them because they modify `self`) and they need to be repeated below (from line 206), I would place them in a function (and call that function from here and below).

Reply:
Unless this will change when we apply the new jitting strategy?